### PR TITLE
Add 'schemars' feature for deriving JSON schema

### PR DESCRIPTION
**Pull request by smklein**
# Context

[schemars](https://docs.rs/schemars/0.8.3/schemars/index.html) is a library which allows emission of JSON schema documents based purely on a type. It's a lot like serde (and has compatibility with serde) but allows emission of JSON schemas without a "value" - it operates purely based on type.

This is super handy for things that consume JSON schemas to define types - at Oxide (my employer), we use schemars to produce OpenAPI specs for our interfaces (and consequently, types that appear in our APIs derive `Serialize`, `Deserialize`, and `JsonSchema`).

Similar to the serde types, `derive(JsonSchema)` only works if all the contained types also implement `JsonSchema`.

# Okay, what about this PR tho

This PR adds a new "feature" to ipnet, enabling the derivation of the `JsonSchema` type for `IpNet` types.
Similar to the existing `serde` feature, a new `json` feature provides a variant of this crate with the derives enabled.

For callers who want to use it: This enables usage of the `IpNet` types within other `JsonSchema` objects.
For callers who don't want it: It's behind a feature flag, so it remains completely unnoticeable (and has no impact on the compiled library).

### CODE REVIEW

**Comment by xfbs** [comment]
+1 on this from me. Although, you can probably get away with not using serde by deriving the JsonSchema yourself.

File: Cargo.toml [hunk 1]
```diff
@@ -10,8 +10,19 @@ categories = ["network-programming"]
 readme = "README.md"
 documentation = "https://docs.rs/ipnet"
 
+[features]
+default = []
+# Implements "serde::Serialize" and "serde::Deserialize".
+serde = ["serde_crate"]
```
Comment (ralpha) [inline Cargo.toml]: Just by making the crate optional (`optional = true`, as already done) you can enable it using a feature flag of the crate name itself.

So then this line is not needed and `serde` feature can just be used immediately. 
For more info: https://stackoverflow.com/a/39759592/2037998

This way some of the changes in `src/ipnet_serde.rs` are no longer needed to. That will make the PR a bit simpler.

Comment (xfbs) [reply Cargo.toml]: I've just pushed something, take a look. Let me know if that is good

Comment (ralpha) [reply Cargo.toml]: I don't see the changes anywhere. Where did you push them?

**Comment by krisprice** [comment]
Thanks @smklein, @ralpha, and @xfbs. Sorry for the long delay. Per the conversation. Shouldn't we be able to just have `schemars` as a new optional dependency here, then without changing anything else we can add the `#[cfg_attr(feature = "json", derive(schemars::JsonSchema))]` lines in `ipnet.rs`? Does that work @smklein ? I'm not sure if the resulting output is satisfactory or if you want it customized a little. If that works and you want to make those changes I'll go ahead and merge this.

**Comment by xfbs** [comment]
Sounds good!

**Comment by ralpha** [comment]
> Shouldn't we be able to just have `schemars` as a new optional dependency here, then without changing anything else we can add the `#[cfg_attr(feature = "json", derive(schemars::JsonSchema))]` lines in `ipnet.rs`?

Yes that would work. But the change in: `src/ipnet_serde.rs:67`
```diff
- match try!(data.variant()) {
+ match data.variant()? {
```
I would keep too. The `try!` macro has been deprecated for some time now: https://doc.rust-lang.org/std/macro.try.html

File: Cargo.toml (hunk 1 above)
Comment (smklein) [reply Cargo.toml]: I went ahead and collapsed this, avoiding the renaming.

`cargo build/test --features serde` and `cargo build/test --features json` are both happy.

**Comment by smklein** [comment]
Avoided making `serde` a separate high-level feature (plus the crate renaming shenanigans) - PTAL!

File: src/lib.rs [hunk 2]
```diff
@@ -79,7 +79,7 @@
 //! [feature]: https://doc.rust-lang.org/cargo/reference/manifest.html#the-features-section
 
 #[cfg(feature = "serde")]
-extern crate serde_crate;
+extern crate serde;
```
Comment (ralpha) [inline src/lib.rs:82]: This code can be removed. This is no longer needed in Rust with `edition = "2018"`, so the edition might have to be set. But can also happen in other PR.

**Comment by ralpha** [comment]
If this compiles and works it looks all good to me.
I guess `#[macro_use]` for serde was never used?

@krisprice I might suggest added an edition to the crate, best to use `edition = "2018"` unless you know that nobody with a Rust version of less then 1.56.0 is using it you can go for 2021. But might be a bit soon for a lower level lib. (can be done separate from this PR)

Other then those remarks it looks all good.

**Comment by krisprice** [comment]
Thanks again @smklein, @ralpha, and @xfbs : now merged and published.

**Pull request merged**